### PR TITLE
ONLINE_BAGGER: Add to sub launch and kill systems

### PR DESCRIPTION
**Pull request by kingkevlar**
Oooooweeeee! Online_bagger has been added to the launch and kill systems for the sub. (launches with the sub and calls the service dump on kill)

Whenever 'araise kill' is called, the current message set is dumped to a bag called 'kill_bag' (see PR in mil_common). It will repeatedly get written over with the name 'kill_bag.bag' to prevent having too many kill bags (this makes it our responsibility to rename bags we care about). I can switch back to the default bag name (current date and time) if we reach some sort of consensus on it. 

I also updated the yaml to bag one camera (right) and the blueview images and ranges by default. (Again, can be changed)

### CODE REVIEW

File: command/sub8_alarm/alarm_handlers/kill.py
Comment (sentree) [inline command/sub8_alarm/alarm_handlers/kill.py]: Try to keep your documentation strings in this format:

    def do_things(self):
    '''
    This function does things.
    ''''
    self.stuff()
    self.things()

Not that there are no spaces between the function definition, documentation string, or first line of code. The only reason for spacing here is if the first line after the documentation string is a comment. I believe this is PEP8, but it may just be OCD; someone correct me if I'm wrong.

File: command/sub8_alarm/alarm_handlers/kill.py
Comment (sentree) [inline command/sub8_alarm/alarm_handlers/kill.py]: By passing bag_time=0, won't we be bagging everything in the cache? Not sure how I feel about this as we may want a larger queue later down the line, but not want to dump the whole thing for a kill.

**Review by sentree — APPROVED** [review]
I saw you and @ironmig freak out about this working, so it was likely well tested.

**Comment by sentree** [comment]
I try to keep bags in a separate partition so they never interfere with our root or home partitions. Also, please make sure your bag ends up in the correct dated folder as defined in the mil_common bag.sh script. Or at least keep in the default bagging folder, which is stored in the `BAG_DIR` environment variable.

File: command/sub8_alarm/alarm_handlers/kill.py
Comment (kingkevlar) [reply command/sub8_alarm/alarm_handlers/kill.py]: I believe PEP8 calls for docstrings in double quotes (which I did not do) and:

 PEP8:
"For one liner docstrings, please keep the closing """ on the same line. "

Comment (sentree) [reply command/sub8_alarm/alarm_handlers/kill.py]: Okay, now I have a beef with PEP8... well, convention is convention!

File: command/sub8_alarm/alarm_handlers/kill.py
Comment (kingkevlar) [reply command/sub8_alarm/alarm_handlers/kill.py]: I didn't think of that, do you think 60 seconds would be an appropriate time?

**Review by sentree — CHANGES_REQUESTED** [review]
On hold until @ironmig's mil_common pull request is merged.

**Comment by kev-the-dev** [comment]
I'm fine with merging once uf-mil/mil_common#48 is merged.

**Comment by sentree** [comment]
Will this integrate well with all of the changes you made?

**Comment by kev-the-dev** [comment]
Doesn't really interface with my changes. But I did pull this branch and test it again uf-mil/mil_common#48 and it seems fine.

File: command/sub8_launch/config/online_bagger.yaml
Comment (kev-the-dev) [inline command/sub8_launch/config/online_bagger.yaml]: This list is unnecessary now that it pulls from the ```bag_*``` environment variables. The bagging topics will now be version controlled through the bag aliases file.

**Review by kev-the-dev — CHANGES_REQUESTED** [review]
One teency redundancy we should get rid of now that onine bagger is integrated with the bag script.

File: command/sub8_alarm/alarm_handlers/kill.py
Comment (sentree) [inline command/sub8_alarm/alarm_handlers/kill.py]: Are we sure we want to dump all topics for the kill bag?

File: command/sub8_launch/launch/subsystems/online_bagger.launch
Comment (sentree) [inline command/sub8_launch/launch/subsystems/online_bagger.launch]: All of @ironmig's shiny new ROS parameters should be added to this launch file. Make sure to leave the bag storage directory blank so it defaults to using the `BAG_DIR` environment variable.

**Review by sentree — COMMENTED** [review]
This is close to ready. It just needs to be updated a bit before it is meged.

File: command/sub8_alarm/alarm_handlers/kill.py
Comment (DSsoto) [reply command/sub8_alarm/alarm_handlers/kill.py]: Hmmm, maybe it should only be the GNC ones and one camera / imaging sonar. Then again maybe we could spare 60 seconds of everything just for kill, not sure how large that gets.

Comment (sentree) [reply command/sub8_alarm/alarm_handlers/kill.py]: Make sure to check next time you guys test. Launch this from @kingkevlar's workspace.

**Comment by DSsoto** [comment]
@kingkevlar, @ironmig, what's the status on this?
I remember seeing you guys fix some bugs but nothing has been pushed here since.

File: command/sub8_alarm/alarm_handlers/kill.py
Comment (mattlangford) [inline command/sub8_alarm/alarm_handlers/kill.py:29]: Have you tried having the sub kill unkill and kill again within a few seconds (or calling the service right after the sub killed)? It looks like there could be a race condition where both requests will try writing to different bags using the same bag object. That may get caught here as an exception but it'll still crash the bag server.

Comment (kingkevlar) [reply command/sub8_alarm/alarm_handlers/kill.py:29]: Do you mean calling the same service simultaneously: as in, while it is being called by the alarm server, it is also getting called in another terminal separately? <- I haven't checked this case, I can check it tomorrow, I didn't think of it.

Or did you mean closely in series as in: run kill (dump bag), then run service again immediately after? <- This one does not cause issues

Comment (mattlangford) [reply command/sub8_alarm/alarm_handlers/kill.py:29]: The former, or just trying to dump a bag while another is being dumped

Comment (DSsoto) [reply command/sub8_alarm/alarm_handlers/kill.py:29]: @kingkevlar, If that's the case, putting a mutex around the `BaggingCommands` function should fix the issue.

Comment (kingkevlar) [reply command/sub8_alarm/alarm_handlers/kill.py:29]: It didn't work when I tried to run them simultaneously, if it was in the middle of a service call, it seems that it would start running the second one in the middle of the first one and the resulting two bags would have the topics split between the two service calls (i.e. I had one bag that had one of the camera topics, and another than had the other half, both had the same amount of messages).

I fixed the issue with a PR to mil_common, if the service is already being called it will reject the call and give priority to the current one. (aka, you cannot run the service in parallel with another service call)

**Comment by kingkevlar** [comment]
@whispercoros It says the commit cannot be built, when I looked at the details it looks like it is related to the thruster_mapper

**Review by kev-the-dev — APPROVED** [review]

File: command/sub8_alarm/alarm_handlers/kill.py
Comment (sentree) [inline command/sub8_alarm/alarm_handlers/kill.py:29]: Everyone remember that we will need to rename the kill bag if we want to save it.

**Review by sentree — APPROVED** [review]
Everything looks good here. Nice job!

**Pull request merged**
